include/rocksdb/wide_columns.h

```diff
@@ -218,4 +224,49 @@ inline bool operator!=(const PinnableWideColumns& lhs,
   return !(lhs == rhs);
 }
 
+inline PinnableWideColumns::PinnableWideColumns(const PinnableWideColumns& p)
```

Comment:
copy constructor was needed for `std::vector` (Is there a way to use the vector of PinnableWideColumns without having copy constructor?)

Reply:
I think "uninitialized copy" is created when `GroupedPinnableWideColumns` is initialized and it uses copy contstructor. The following is the stack when the copy constructor was hit.

```
rocksdb::PinnableWideColumns::PinnableWideColumns(rocksdb::PinnableWideColumns const&) (/data/users/jewoongh/rocksdb/include/rocksdb/wide_columns.h:227)
void std::_Construct<rocksdb::PinnableWideColumns, rocksdb::PinnableWideColumns const&>(rocksdb::PinnableWideColumns*, rocksdb::PinnableWideColumns const&) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_construct.h:109)
rocksdb::PinnableWideColumns* std::__uninitialized_copy<false>::__uninit_copy<__gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, rocksdb::PinnableWideColumns*>(__gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, __gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, rocksdb::PinnableWideColumns*) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_uninitialized.h:92)
rocksdb::PinnableWideColumns* std::uninitialized_copy<__gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, rocksdb::PinnableWideColumns*>(__gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, __gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, rocksdb::PinnableWideColumns*) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_uninitialized.h:151)
rocksdb::PinnableWideColumns* std::__uninitialized_copy_a<__gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, rocksdb::PinnableWideColumns*, rocksdb::PinnableWideColumns>(__gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, __gnu_cxx::__normal_iterator<rocksdb::PinnableWideColumns const*, std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>>, rocksdb::PinnableWideColumns*, std::allocator<rocksdb::PinnableWideColumns>&) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_uninitialized.h:333)
std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>>::vector(std::vector<rocksdb::PinnableWideColumns, std::allocator<rocksdb::PinnableWideColumns>> const&) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_vector.h:558)
rocksdb::GroupedPinnableWideColumns::GroupedPinnableWideColumns(rocksdb::GroupedPinnableWideColumns const&) (/data/users/jewoongh/rocksdb/include/rocksdb/wide_columns.h:233)
void std::_Construct<rocksdb::GroupedPinnableWideColumns, rocksdb::GroupedPinnableWideColumns const&>(rocksdb::GroupedPinnableWideColumns*, rocksdb::GroupedPinnableWideColumns const&) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_construct.h:109)
rocksdb::GroupedPinnableWideColumns* std::__uninitialized_copy<false>::__uninit_copy<rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns*>(rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns*) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_uninitialized.h:92)
rocksdb::GroupedPinnableWideColumns* std::uninitialized_copy<rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns*>(rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns*) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_uninitialized.h:151)
rocksdb::GroupedPinnableWideColumns* std::__uninitialized_copy_a<rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns*, rocksdb::GroupedPinnableWideColumns>(rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns*, std::allocator<rocksdb::GroupedPinnableWideColumns>&) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_uninitialized.h:333)
void std::vector<rocksdb::GroupedPinnableWideColumns, std::allocator<rocksdb::GroupedPinnableWideColumns>>::_M_range_initialize<rocksdb::GroupedPinnableWideColumns const*>(rocksdb::GroupedPinnableWideColumns const*, rocksdb::GroupedPinnableWideColumns const*, std::forward_iterator_tag) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_vector.h:1585)
std::vector<rocksdb::GroupedPinnableWideColumns, std::allocator<rocksdb::GroupedPinnableWideColumns>>::vector(std::initializer_list<rocksdb::GroupedPinnableWideColumns>, std::allocator<rocksdb::GroupedPinnableWideColumns> const&) (/mnt/gvfs/third-party2/libgcc/c00dcc6a3e4125c7e8b248e9a79c14b78ac9e0ca/11.x/platform010/5684a5a/include/c++/trunk/bits/stl_vector.h:629)
```

Reply:
>are we copying a vector of PWCs somewhere?

Yeah, I think the copy is happening when I do the following in the test code

```
...
std::vector<GroupedPinnableWideColumns> results{first_key_result, second_key_result};
...
```
If my understanding is correct, each element is copy-initialized here. It calls the copy constructor of `GroupedPinnableWideColumns`, and the default copy constructor of `GroupedPinnableWideColumns` is copying `std::vector<PinnableWideColumns>`.

Let me see what I can do to do this without adding copy constructor to PWC

Reply:
You could either move `first_key_result` and `second_key_result` into the container or alternatively, pass temporaries instead of named variables to the vector constructor.